libchisel/src/remapstart.rs

```diff
@@ -0,0 +1,228 @@
+            .entries_mut()
+            .iter_mut()
+        {
+            if *export_entry.field() == "main".to_string() {
+                *export_entry =
+                    ExportEntry::new("main".to_string(), Internal::Function(start_func_idx));
+                ret = true;
+            }
+        }
+    }
+
+    return ret;
+}
+
+// if the main export refers to a different function than "main" from the name section
+// rename the main function referenced by the name section by prepending one or more '_'
+// add the main export to the name section (?)
+fn rename_main(module: &mut Module) {
+    let start_func_idx = module.start_section().expect("no start section");
+    let mut export_name = "_main".to_string();
+    let export_main_func_idx = module
+        .export_section()
+        .expect("no export section")
+        .entries()
+        .iter()
+        .find_map(|s| match s.internal() {
+            Internal::Function(f) => {
+                if s.field() == "main".to_string() {
+                    Some(f)
+                } else {
+                    None
+                }
+            }
+            _ => None,
+        })
+        .expect("main not exported")
+        .clone();
+
+    let try_replace =
+        |module: &mut Module, func_name: &str, func_idx: u32, main_func_idx: u32| -> bool {
+            match module.names_section_mut() {
+                Some(NameSection::Function(function_name_section)) => {
+                    match function_name_section
+                        .names()
+                        .iter()
+                        .find(|(_, s)| **s == func_name)
+                    {
+                        Some(_) => false,
+                        None => {
+                            function_name_section
+                                .names_mut()
+                                .insert(func_idx, func_name.to_string());
+
+                            function_name_section
+                                .names_mut()
+                                .insert(main_func_idx, "main".to_string());
+
+                            true
+                        }
+                    }
+                }
+                Some(_) => {
+                    panic!("names_section_mut should return a function name section");
+                }
+                None => {
+                    panic!("name section not found");
+                }
+            }
+        };
+
+    while !try_replace(module, &export_name, start_func_idx, export_main_func_idx) {
+        export_name = format!("_{}", export_name);
+    }
+}
+
+fn remap_start(module: &mut Module, translations: &RemapStart) -> bool {
+    let mut ret = replace_main_export(module);
+
+    if ret {
+        let name_section_exists: bool =
+            if let Some(NameSection::Function(_)) = module.names_section() {
+                true
+            } else {
+                false
+            };
+
+        if name_section_exists {
```

Comment:
You can replace this entire blob with `module.names_section().is_some()`.

Reply:
I still don't get why you need the names section to rename exported symbols. Those are not part of the names section.